File: pRobot169/perl/run_daily.pl
#!/usr/bin/perl
#

$usage = <<USAGE;
Usage:
  perl $0 exeGetList uid_pass.txt
  exeGetList:      the perl script to get article list
  uid_pass.txt:    uid & password file, select non-existent file means
                   not to webpost
USAGE

$exeGetList = shift or die $usage;
-f $exeGetList or die "$exeGetList not exists!\n$usage";
$fileUidPasswd = shift or die $usage;
-f $fileUidPasswd or print "$fileUidPasswd not found -> will not post\n";

my @fileBaha = ();
my $outfileBaha;
&refreshFileBaha;

my %do_at;
while (1) {
    my $hrmin = &date("+%H%M");
    my $hr = substr($hrmin,0,2);
    my $min = substr($hrmin,2,2);
    last if $hrmin le "0300";

    if (($hr =~ /^(06|18|22)$/) && ! defined $do_at{$hr."00"}) {
        &runGetList($exeGetList);
        $do_at{$hr."00"} = 1;
    }
    elsif (($hrmin gt "2330") && ! defined $do_at{"2330"}) {
        &runGetList($exeGetList);
        $do_at{"2330"} = 1;
    }
    elsif ($hrmin gt "0700") {
        &refreshFileBaha;
        if (0 == scalar(@fileBaha)) {
            &runGetList($exeGetList);
        }
    }

    my $hrmin = &date("+%H%M");
    $hr = substr($hrmin,0,2);
    $min = substr($hrmin,2,2);
    my $minWait = 10;

    last if $hrmin le "0300";
    if ($hrmin le "0600") {
        $minWait = 60;
    }
    elsif ($hrmin gt "2330") {
        $minWait = 1;
    }

    my $secWait = $minWait*60;
    printf("%s - sleep for %d secs\n", &date, $secWait);
    sleep $secWait;
}

# --- get the year / month / mday ------------------
($p2d_yr, $p2d_mon, $p2d_mday,
    $yesterday_yr, $yesterday_mon, $yesterday_mday,
    $today_yr, $today_mon, $today_mday) = split(/ /, `perl dates.pl`);

$yesterday = "$yesterday_mon/$yesterday_mday";
$yesterday8 = "$yesterday_yr$yesterday_mon$yesterday_mday";
-d $yesterday8 && exit;
# ==================================================

print "Call last1.pl...\n";
system("perl", "last1.pl", $exeGetList,
    $p2d_yr, $p2d_mon, $p2d_mday,
    $yesterday_yr, $yesterday_mon, $yesterday_mday,
    $today_yr, $today_mon, $today_mday);

if (-f $fileUidPasswd) { 
    print "Posting...\n";
    &cp("$yesterday8/stat.txt", "stat.txt");
    system("perl webpost07.pl Chat $yesterday $fileUidPasswd");
    unlink "stat.txt";
}
exit;

sub refreshFileBaha {
    @fileBaha = ();
    $outfileBaha = undef;
    for my $i (1 .. 99) {
        my $tFileBaha = sprintf("baha%02d.txt", $i);
        if (-f $tFileBaha) {
            push @fileBaha, $tFileBaha;
        }
        elsif (! defined($outfileBaha)) {
            $outfileBaha = $tFileBaha;
        }
    }
}

sub runGetList {
    system("perl", @_);
    &refreshFileBaha;
}

sub CSTtime () {
    return time+8*3600;
}

use POSIX qw(strftime);
sub date ($) {
    my $strFormat = shift || "%a %b %e %H:%M:%S %Y";
    $strFormat =~ s/^\+//g;
    return strftime($strFormat, gmtime(&CSTtime));
}

sub cp ($$) {
    my ($from, $to) = @_;
    my ($buf);
    
    open(FROM, "$from") || die "Can't open $from\n";
    open(TO, ">$to")    || die "Can't open $to\n";

    while(read (FROM, $buf, 16384)) {
        print TO $buf;
    }
    close(FROM);
    close(TO);
}

File: pRobot169/python/run_daily.py
import os
import sys
import time
import shutil
from datetime import datetime


def refreshFileBaha():
    global fileBaha, outfileBaha
    fileBaha = []
    outfileBaha = None
    for i in range(1, 100):
        tFileBaha = f"baha{i:02d}.txt"
        if os.path.isfile(tFileBaha):
            fileBaha.append(tFileBaha)
        elif outfileBaha is None:
            outfileBaha = tFileBaha

def runGetList(exeGetList):
    os.system(f"python {exeGetList}")
    refreshFileBaha()

def CSTtime():
    return time.time() + 8 * 3600

def date(strFormat="%a %b %e %H:%M:%S %Y"):
    strFormat = strFormat.lstrip("+")
    return datetime.fromtimestamp(CSTtime()).strftime(strFormat)

def cp(from_path, to_path):
    shutil.copyfile(from_path, to_path)

usage = """Usage:
  python {0} exeGetList uid_pass.txt
  exeGetList:      the python script to get article list
  uid_pass.txt:    uid & password file, select non-existent file means
                   not to webpost
""".format(sys.argv[0])

exeGetList = sys.argv[1] if len(sys.argv) > 1 else (print(usage) or sys.exit(1))
if not os.path.isfile(exeGetList):
    print(f"{exeGetList} not exists!\n{usage}")
    sys.exit(1)

fileUidPasswd = sys.argv[2] if len(sys.argv) > 2 else (print(usage) or sys.exit(1))
if not os.path.isfile(fileUidPasswd):
    print(f"{fileUidPasswd} not found -> will not post\n")

fileBaha = []
outfileBaha = None
refreshFileBaha()

do_at = {}
while True:
    hrmin = datetime.now().strftime("%H%M")
    hr = hrmin[:2]
    min = hrmin[2:]
    if hrmin <= "0300":
        break

    if hr in ["06", "18", "22"] and not do_at.get(f"{hr}00"):
        runGetList(exeGetList)
        do_at[f"{hr}00"] = 1
    elif hrmin > "2330" and not do_at.get("2330"):
        runGetList(exeGetList)
        do_at["2330"] = 1
    elif hrmin > "0700":
        refreshFileBaha()
        if len(fileBaha) == 0:
            runGetList(exeGetList)

    hrmin = datetime.now().strftime("%H%M")
    hr = hrmin[:2]
    min = hrmin[2:]
    minWait = 10

    if hrmin <= "0300":
        break
    if hrmin <= "0600":
        minWait = 60
    elif hrmin > "2330":
        minWait = 1

    secWait = minWait * 60
    print(f"{datetime.now()} - sleep for {secWait} secs")
    time.sleep(secWait)

p2d_yr, p2d_mon, p2d_mday, yesterday_yr, yesterday_mon, yesterday_mday, today_yr, today_mon, today_mday = os.popen("python dates.py").read().split()

yesterday = f"{yesterday_mon}/{yesterday_mday}"
yesterday8 = f"{yesterday_yr}{yesterday_mon}{yesterday_mday}"
if os.path.isdir(yesterday8):
    sys.exit()

print("Call last1.py...")
os.system(f"python last1.py {exeGetList} {p2d_yr} {p2d_mon} {p2d_mday} {yesterday_yr} {yesterday_mon} {yesterday_mday} {today_yr} {today_mon} {today_mday}")

if os.path.isfile(fileUidPasswd):
    print("Posting...")
    shutil.copyfile(f"{yesterday8}/stat.txt", "stat.txt")
    os.system(f"python webpost07.py Chat {yesterday} {fileUidPasswd}")
    os.remove("stat.txt")

sys.exit()

